scipy/signal/_spectral_py.py

```diff
@@ -240,81 +244,106 @@ def lombscargle(
     # weight vector must sum to 1
     weights = weights / weights.sum()
 
-    # pre-allocate arrays for intermediate and final calculations
-    coswt = np.empty_like(x)
-    sinwt = np.empty_like(x)
-    pgram = np.empty_like(freqs)
-    # store a and b so that phase can be calculated outside loop, if necessary
-    a = np.empty_like(freqs)
-    b = np.empty_like(freqs)
-
     # if requested, perform precenter
     if precenter:
         y -= y.mean()
 
-    # calculate the single sum that does not depend on the frequency
-    Y_sum = (weights * y).sum()
+    # transform arrays
+    # row vector
+    freqs = freqs[np.newaxis, :]
+    # column vectors
+    x = x[:, np.newaxis]
+    y = y[:, np.newaxis]
+    weights = weights[:, np.newaxis]  # type: ignore
 
-    # loop over the frequencies
-    for i in range(freqs.shape[0]):
-        coswt[:] = np.cos(freqs[i] * x)
-        sinwt[:] = np.sin(freqs[i] * x)
+    weights_y = weights * y
+    Y_sum = (weights_y).sum()
 
-        # calculate CC, SS, and CS first
-        # these variable names are of form XX_hat in the paper, but dropping the suffix
-        # to reuse variables later whether floating_mean is True or False
-        CC = (weights * coswt * coswt).sum()
-        SS = 1 - CC  # trig identity: S^2 = 1 - C^2
-        CS = (weights * coswt * sinwt).sum()
+    freqst = freqs * x
+    coswt = np.cos(freqst)
+    sinwt = np.sin(freqst)
 
-        # now, redefine trig arrays to always be multiplied by the weights
-        coswt[:] = weights * coswt
-        sinwt[:] = weights * sinwt
+    CC = 2.0 * np.dot(weights.T, 0.5 - sinwt**2.0)
+    SS = 2.0 * np.dot(weights.T, coswt * sinwt)
 
-        # these are also of the form XX_hat in the paper
-        YC = (y * coswt).sum()
-        YS = (y * sinwt).sum()
+    if floating_mean:
+        C = np.dot(weights.T, coswt)
+        S = np.dot(weights.T, sinwt)
+        CC -= C * C - S * S
+        SS -= 2.0 * C * S
 
-        if floating_mean:
-            # calculate best-fit offset for each frequency independently (default)
-            C_sum = coswt.sum()
-            S_sum = sinwt.sum()
-            YC -= Y_sum * C_sum
-            YS -= Y_sum * S_sum
-            CC -= C_sum * C_sum
-            SS -= S_sum * S_sum
-            CS -= C_sum * S_sum
-
-        # determinate of the system of linear equations
-        D = CC * SS - CS * CS
-
-        # where: y(w) = a*cos(w) + b*sin(w) + c
-        a[i] = (YC * SS - YS * CS) / D
-        b[i] = (YS * CC - YC * CS) / D
-        #  c = Y_sum - a * C_sum - b * S_sum  # offset is not useful to return
-
-        # store final value as power in (y units)^2
-        pgram[i] = 2.0 * (a[i] * YC + b[i] * YS)
-
-    if normalize == "normalize":
+    # calculate tau (phase offset to eliminate CS variable)
+    tau = 0.5 * np.arctan2(SS, CC)
+    freqst_tau = freqst - tau
```

Comment:
Yeah, the tau parameter is in all 4 papers that are referenced in the docstring. For various reasons I haven't always used the trick (as it isn't strictly necessary). But I was digging deep looking for a better solution, and realized getting rid of CS was possibly the right route.

Reply:
I decided to write the tau equation so that it matches the standard form (instead of the trig-identity simplified one). And since I was already pointing to the paper, I decided to label where every equation came from. I can remove those labels though if desired.